fix(elements): update the view of an `OnPush` component when inputs change

gkalpak opened the pull request:
As with regular Angular components, Angular elements are expected to have their views update when inputs change.

Previously, Angular Elements views where not update if the underlying component used the `OnPush` change detection strategy.

This commit fixes this by calling `markForCheck()` on the component view's `ChangeDetectorRef`.

NOTE:
This is similar to how `@angular/upgrade` does it: https://github.com/angular/angular/blob/3236ae0ee118d0734c90fa9f3767435396213470/packages/upgrade/src/common/src/downgrade_component_adapter.ts#L146.

Fixes #38948.

##
This PR also contains some commit the fix/improve the `elements` tests (both unit and e2e tests). See individual commits for more details.

mary-poppins commented:
You can preview 02a11b6 at https://pr39452-02a11b6.ngbuilds.io/.

mary-poppins commented:
You can preview 44dc930 at https://pr39452-44dc930.ngbuilds.io/.

mary-poppins commented:
You can preview ed44a83 at https://pr39452-ed44a83.ngbuilds.io/.

gkalpak commented on goldens/size-tracking/aio-payloads.json:
```diff
@@ -12,7 +12,7 @@
       "uncompressed": {
         "runtime-es2015": 3037,
-        "main-es2015": 448085,
+        "main-es2015": 448756,
```
The actual increase here is 277B (from [448479][1] to [448756][2]).

[1]: https://app.circleci.com/pipelines/github/angular/angular/23665/workflows/48f3734b-5ed8-43d8-9b50-6c3bf0d21186/jobs/842748
[2]: https://app.circleci.com/pipelines/github/angular/angular/23709/workflows/82d2e0ea-a037-4ea3-a81d-edb625870cf4/jobs/843518

gkalpak commented on goldens/size-tracking/aio-payloads.json:
```diff
@@ -21,7 +21,7 @@
       "uncompressed": {
         "runtime-es2015": 3157,
-        "main-es2015": 431750,
+        "main-es2015": 432151,
```
The actual increase here is 277B (from [431874][1] to [432151][2]).

[1]: https://app.circleci.com/pipelines/github/angular/angular/23665/workflows/48f3734b-5ed8-43d8-9b50-6c3bf0d21186/jobs/842746
[2]: https://app.circleci.com/pipelines/github/angular/angular/23709/workflows/82d2e0ea-a037-4ea3-a81d-edb625870cf4/jobs/843515

josephperrott reviewed (approved):
LGTM

petebacondarwin reviewed (commented):
typo in the last commit:

`where not update` -> `were not updated`

petebacondarwin reviewed (approved):
LGTM 

Reviewed-for: size-tracking

mary-poppins commented:
You can preview 68ab821 at https://pr39452-68ab821.ngbuilds.io/.

IgorMinar reviewed (approved):
LGTM 

Reviewed-for: size-tracking

gkalpak commented on packages/elements/test/component-factory-strategy_spec.ts:
```diff
@@ -261,6 +264,55 @@ describe('ComponentFactoryNgElementStrategy', () => {
+         expect(viewChangeDetectorRef.markForCheck).toHaveBeenCalledTimes(1);
+       }));
+
+    it('should mark the view for check twice for changes in different rounds with previous values',
```
Good questions. I've added tests for the first case (setting to the same value in a new round).

For the second case (setting to a new value and then back to the original value), I don't think we'll gain much by adding tests. There is no logic to detect that, so `detectChanges()` (including `markForCheck()`) will run (although redundant 🤷‍♂️).

mary-poppins commented:
You can preview 52d608e at https://pr39452-52d608e.ngbuilds.io/.

gkalpak commented:
NOTE: #39463 is trying to address the issue more generically.

gkalpak committed:
test(elements): fix typo in test description (`even if when` --> `even if`)

This commit fixes a typo in a test description.

gkalpak committed:
test(elements): fix `ComponentNgElementStrategy` test for components without `ngOnChanges`

`ComponentNgElementStrategy` is supposed to call `ngOnChanges()` on the
underlying component instance if available, but not fail if the
component does not have an `ngOnChanges()` method. This works as
expected. However, the test used to verify that was invalid; i.e. the
test would pass even if `ComponentNgElementStrategy` would try to call
`ngOnChanges()` on a component without such a method.

This commit replaces the invalid test with a new one that correctly
verifies that `ComponentNgElementStrategy` does not try to call
`ngOnChanges()`.

gkalpak committed:
test(elements): simplify creation of `FakeComponentFactory#componentRef`

Previously, the `componentRef` property of `FakeComponentFactory` used
in `elements` tests was initialy set to a spy object with all mock
properties defined as spied methods. Later, the properties where
overwritten to the actual mock values.

This commit simplifies the creation of `componentRef` by correctly using
the arguments of [jasmine.createSpyObj()][1] to specify the desired
shape of the spy object (separating spied properties from methods and
directly providing the mock values).

[1]: https://jasmine.github.io/api/3.5/jasmine.html#.createSpyObj

gkalpak committed:
test(elements): add integration tests for Angular Elements using `ShadowDom`

Previously, the project used for running integration tests for Angular
Elements declared a component that used `ShadowDom` for view
encopsulation, but it did not include any tests to verify that the view
was updated correctly.

This commit adds the missing tests.

gkalpak committed:
fix(elements): update the view of an `OnPush` component when inputs change

As with regular Angular components, Angular elements are expected to
have their views update when inputs change.

Previously, Angular Elements views were not updated if the underlying
component used the `OnPush` change detection strategy.

This commit fixes this by calling `markForCheck()` on the component
view's `ChangeDetectorRef`.

NOTE:
This is similar to how `@angular/upgrade` does it:
https://github.com/angular/angular/blob/3236ae0ee118d0734c90fa9f3767435396213470/packages/upgrade/src/common/src/downgrade_component_adapter.ts#L146.

Fixes #38948

gkalpak committed:
fixup! fix(elements): update the view of an `OnPush` component when inputs change

mary-poppins commented:
You can preview dd1ee40 at https://pr39452-dd1ee40.ngbuilds.io/.

gkalpak commented on goldens/size-tracking/aio-payloads.json:
```diff
@@ -12,7 +12,7 @@
       "uncompressed": {
         "runtime-es2015": 3037,
-        "main-es2015": 448615,
+        "main-es2015": 448922,
```
The actual increase here is 277B (from [448645][1] to [448922][2]).

[1]: https://app.circleci.com/pipelines/github/angular/angular/24192/workflows/1f86d952-a813-4e00-8aba-11c2420c18dc/jobs/850932
[2]: https://app.circleci.com/pipelines/github/angular/angular/24189/workflows/03d37062-e9d7-401e-a563-7a386db54994/jobs/850877

gkalpak commented on goldens/size-tracking/aio-payloads.json:
```diff
@@ -21,7 +21,7 @@
       "uncompressed": {
         "runtime-es2015": 3157,
-        "main-es2015": 431750,
+        "main-es2015": 432199,
```
The actual increase here is 277B (from [431922][1] to [432199][2]).

[1]: https://app.circleci.com/pipelines/github/angular/angular/24192/workflows/1f86d952-a813-4e00-8aba-11c2420c18dc/jobs/850934
[2]: https://app.circleci.com/pipelines/github/angular/angular/24189/workflows/03d37062-e9d7-401e-a563-7a386db54994/jobs/850878

mary-poppins commented:
You can preview ad13d8b at https://pr39452-ad13d8b.ngbuilds.io/.

atscott commented:
[global presubmit](https://test.corp.google.com/ui#id=OCL:340874556:BASE:340874705:1604599375421:32d7e440)
